Consider onion services secure for cookies

jtracey opened the pull request:
### Summary

<!-- Provide a general description of the code changes in your pull
request... were there any bugs you had fixed? If so, mention them. If
these bugs have open GitHub issues, be sure to tag them here as well,
to keep the conversation linked together. -->
Adds a new `trustworthy?` method to the `Request` class, and uses it to consider HTTPS as well as onion services to be secure for the purposes of the `secure` flag in cookies. Fixes #41426.

### Other Information

I added the `trustworthy?` method instead of directly checking in `write_cookie?` for onion services, since this allows any other spots in the code that should be conditional on secure connections rather than HTTPS specifically to use it instead (anything we're missing now, plus any future code), and it gives a single place to add any future fancy secure standards that browsers implement.

The name "trustworthy?" comes from the [W3C secure contexts draft](https://w3c.github.io/webappsec-secure-contexts/#is-origin-trustworthy), which is roughly [what Firefox uses to decide whether secure cookies should be sent](https://bugzilla.mozilla.org/show_bug.cgi?id=1633015). It's a bit of a misnomer though, since one of the main things the draft lists as "potentially trustworthy origins" is localhost, which Rails does not. I would be willing to change the name to something less misleading (e.g., "secure?"), or to add localhost connections to the list of connections considered trustworthy, but the latter would be a pretty major change in how Rails works IIUC, so I thought I'd leave it for others to decide.

Of course, let me know anything that should be different (names, structure, tests, etc.).

jtracey commented:
Whoops, forgot to tag: Fixes #41426

jtracey commented:
I closed this because I had a moment of panic that an active network adversary could modify an unsecured HTTP connection so that the host field was a `.onion` address, while still requesting the same resource, and a default configuration running a single host would accept that and consider it secure. I thought about ways to solve this, then remembered that such an adversary could also just turn an HTTP connection into an actual HTTPS connection, and accomplish the same thing with better results, so this actually isn't a new problem (i.e., unsecured HTTP is going to be unsecured, we don't need to worry about that case).

jtracey commented:
Is there something else I should be doing to get this merged?

zzak commented:
@jtracey Thank you for your patience, hopefully someone will be able to review this soon 🙏

chiragsinghal commented on actionpack/lib/action_dispatch/http/request.rb:
```diff
@@ -438,6 +438,10 @@ def ssl?
     end
 
+    def trustworthy?
+      ssl? || host.end_with?(".onion")
```
instead of doing this as a global change for all Rails apps, would it make sense to expose this as a config so that relevant apps can define one or more trusted domains?

But then, `.onion` is a special use domain, so doing this as a global change might be fine too.

zzak commented:
@jtracey Would you consider making this behavior optional (behind a setting)? Also looks like it needs a rebase :pray:

jtracey commented:
@chiragsinghal @zzak I'd much rather leave .onion as a secure context by default. From an administrator perspective, onion services tend to be pretty close to 0 configuration—just point Tor at the port the server is listening on, and it works the same as any other service. Adding an "Oh, by the way, if you're using a Rails app, you should also do this" instruction would be unfortunate. From a standards perspective, the [IETF](https://datatracker.ietf.org/doc/html/rfc7686) considers .onion to signify a secure, authenticated connection. The two major browsers that support .onion addresses by default, Brave and Tor Browser, treat it as a secure connection, as does Firefox, which doesn't support connecting to .onion addresses by default, but has plans to (and can of course be configured to do so now). If major supported clients all expect a certain behavior, particularly as part of an open standard backed by large players like Facebook, but Rails does not comply with that expected behavior without additional configuration, I would consider that a bug. I'm amenable to allowing this to be customized, but unless there's a particular reason not to, I'd like to leave .onion in the default trustworthy TLDs.


>Also looks like it needs a rebase

Sure thing.  Is rebase + force push, or a merge commit preferred?

rafaelfranca commented on actionpack/lib/action_dispatch/http/request.rb:
```diff
@@ -438,6 +438,10 @@ def ssl?
       super || scheme == "wss"
     end
 
+    def trustworthy?
```
Since this method is only used in the cookies, let's move it there. There is no need to expose it in the request.

jtracey commented:
I've rebased, and rewritten to narrow the scope of the change, as requested. It now just does an additional check for a .onion domain when sending secure cookies. I was originally hoping to make it more extensible for future changes in what protocols are considered "trustworthy", but it does make sense that future-proofing when the current requirements are so simple is a bit premature.

jtracey committed:
Consider onion services secure for cookies
